src/ctap/mod.rs

```diff
@@ -244,6 +244,123 @@ fn verify_signature(
+        match ctap_hid_channel.send_or_recv_with_timeout(&mut pkt, timeout) {
+            Ok(SendOrRecvStatus::Timeout) => {
+                debug_ctap!(env, "Sending a KEEPALIVE packet timed out");
+                // TODO: abort user presence test?
+            }
+            Err(_) => panic!("Error sending KEEPALIVE packet"),
+            Ok(SendOrRecvStatus::Sent) => {
+                debug_ctap!(env, "Sent KEEPALIVE packet");
+            }
+            Ok(SendOrRecvStatus::Received) => {
+                // We only parse one packet, because we only care about CANCEL.
+                let (received_cid, processed_packet) = CtapHid::process_single_packet(&pkt);
+                if received_cid != &cid {
+                    debug_ctap!(
+                        env,
+                        "Received a packet on channel ID {:?} while sending a KEEPALIVE packet",
+                        received_cid,
+                    );
+                    return Ok(());
+                }
+                match processed_packet {
+                    ProcessedPacket::InitPacket { cmd, .. } => {
+                        if cmd == CtapHidCommand::Cancel as u8 {
+                            // We ignore the payload, we can't answer with an error code anyway.
+                            debug_ctap!(env, "User presence check cancelled");
+                            return Err(Ctap2StatusCode::CTAP2_ERR_KEEPALIVE_CANCEL);
+                        } else {
+                            debug_ctap!(
+                                env,
+                                "Discarded packet with command {} received while sending a KEEPALIVE packet",
+                                cmd,
+                            );
+                        }
+                    }
+                    ProcessedPacket::ContinuationPacket { .. } => {
+                        debug_ctap!(
+                            env,
+                            "Discarded continuation packet received while sending a KEEPALIVE packet",
+                        );
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Blocks for user presence.
+///
+/// Returns an error in case of timeout, user declining presence request, or keepalive error.
+pub fn check_user_presence(env: &mut impl Env, channel: Channel) -> Result<(), Ctap2StatusCode> {
+    if let Some(result) = env.user_presence().quick_check(channel) {
+        return result;
+    }
+
+    env.user_presence().user_presence_check_init(channel);
+
+    // The timeout is N times the keepalive delay.
+    const TIMEOUT_ITERATIONS: usize = TOUCH_TIMEOUT_MS as usize / KEEPALIVE_DELAY_MS as usize;
+
+    // First, send a keep-alive packet to notify that the keep-alive status has changed.
+    // All fallible functions are called without '?' operator to always reach
+    // user_presence_check_complete(...) cleanup function.
+    let mut result = send_keepalive_up_needed(env, channel, KEEPALIVE_DELAY)
```

Comment:
Can't this whole code look like the following instead?

```rust
check_init();
let mut result = ...;
for i in 0..=TIMEOUT_ITERATIONS {
    result = send_keepalive();
    if result.is_err() || i == TIMEOUT_ITERATIONS { break; }
    result = wait_with_timeout();
    if result.is_err() { break; }
}
check_complete();
```
        
Essentially having only one call-site for `send_keepalive`. Unless I'm mistaken this should have the same behavior.